### PR TITLE
Update vSphere  CSI driver to v2.7.0

### LINKED ISSUE #8324: Update vSphere CSI addon
**User Story**

The latest version of the vSphere CSI driver supports Kubernetes 1.22 clusters. We should update to reintroduce 1.22 support for vSphere clusters.

https://github.com/kubernetes-sigs/vsphere-csi-driver/releases/tag/v2.4.0

**Acceptance criteria**

* vSphere CSI addon is updated to v2.4.0
* Kubernetes 1.22 clusters is supported on vSphere

### LINKED ISSUE #10461: Upgrade vsphere CSI to 2.6.0 for upstream bug fix
### Description of the feature you would like to add / User story
At our customer, we are facing [an issue with vsphere volumes](https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/1636) which is fixed in [vsphere CSI driver v2.6.0](https://docs.vmware.com/en/VMware-vSphere-Container-Storage-Plug-in/2.6/rn/vmware-vsphere-container-storage-plugin-26-release-notes/index.html)

This version is [now available in kubeone](https://github.com/kubermatic/kubeone/pull/2169) as well

So request here is to make this CSI driver addon available in KKP as well for user-cluster deployments.

### LINKED ISSUE #10719: vsphere csi addon - vmware-system-csi namespace mandate
### What happened?
Currently, vsphere-csi addon gets deployed in `kube-system` namespace. As per https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/615 - it is required to be vmware-system-csi namespace.

We are currently facing some errors in volume attachment and there a very specific error is being thrown back..

```
{"level":"warn","time":"2022-08-09T11:44:58.191023507Z","caller":"volume/manager.go:319","msg":"failed to store CreateVolume details with error: namespaces \"vmware-system-csi\" not found","TraceId":"1272b390-0c6b-473e-95ea-11e8eded394e"}

{"level":"error","time":"2022-08-09T11:44:58.190708545Z","caller":"cnsvolumeoperationrequest/cnsvolumeoperationrequest.go:196","msg":"failed to create CnsVolumeOperationRequest instance vmware-system-csi/b1ad9273-17d8-11ed-b5d8-c6ddef242ceb with error: namespaces \"vmware-system-csi\" not found","TraceId":"1272b390-0c6b-473e-95ea-11e8eded394e","stacktrace":"sigs.k8s.io/vsphere-csi-driver/v2/pkg/internalapis/cnsvolumeoperationrequest.(*operationRequestStore).StoreRequestDetails
    /build/pkg/internalapis/cnsvolumeoperationrequest/cnsvolumeoperationrequest.go:196
sigs.k8s.io/vsphere-csi-driver/v2/pkg/common/cns-lib/volume.(*defaultManager).createVolumeWithImprovedIdempotency.func1
    /build/pkg/common/cns-lib/volume/manager.go:317
sigs.k8s.io/vsphere-csi-driver/v2/pkg/common/cns-lib/volume.(*defaultManager).createVolumeWithImprovedIdempotency
    /build/pkg/common/cns-lib/volume/manager.go:434
sigs.k8s.io/vsphere-csi-driver/v2/pkg/common/cns-lib/volume.(*defaultManager).CreateVolume.func1
    /build/pkg/common/cns-lib/volume/manager.go:567
sigs.k8s.io/vsphere-csi-driver/v2/pkg/common/cns-lib/volume.(*defaultManager).CreateVolume
    /build/pkg/common/cns-lib/volume/manager.go:572
sigs.k8s.io/vsphere-csi-driver/v2/pkg/apis/migration.(*volumeMigration).registerVolume
    /build/pkg/apis/migration/migration.go:476
sigs.k8s.io/vsphere-csi-driver/v2/pkg/apis/migration.(*volumeMigration).GetVolumeID
    /build/pkg/apis/migration/migration.go:211
sigs.k8s.io/vsphere-csi-driver/v2/pkg/csi/service/vanilla.(*controller).ControllerUnpublishVolume.func1
    /build/pkg/csi/service/vanilla/controller.go:1141
sigs.k8s.io/vsphere-csi-driver/v2/pkg/csi/service/vanilla.(*controller).ControllerUnpublishVolume
    /build/pkg/csi/service/vanilla/controller.go:1162
github.com/container-storage-interface/spec/lib/go/csi._Controller_ControllerUnpublishVolume_Handler
    /go/pkg/mod/github.com/container-storage-interface/spec@v1.4.0/lib/go/csi/csi.pb.go:5643
google.golang.org/grpc.(*Server).processUnaryRPC
    /go/pkg/mod/google.golang.org/grpc@v1.27.1/server.go:1024
google.golang.org/grpc.(*Server).handleStream
    /go/pkg/mod/google.golang.org/grpc@v1.27.1/server.go:1313
google.golang.org/grpc.(*Server).serveStreams.func1.1
    /go/pkg/mod/google.golang.org/grpc@v1.27.1/server.go:722"}
```

### Expected behavior
vsphere-csi-driver addon should get deployed in it's own dedicated namesapce and volume attach should succeed.

### How to reproduce the issue?
Unfortunatly, right now, there are no confirm steps to reproduce the issue. It started appearing in our clusters out of blue

### How is your environment configured?

- KKP version: 2.19.8
- Shared or separate master/seed clusters?: separate

### Provide your KKP manifest here (if applicable)

<details>

```yaml
# paste manifest here
```

</details>

### What cloud provider are you running on?

Vsphere

### What operating system are you running in your user cluster?

Ubuntu 20.04

### PR DESCRIPTION
**Pull request by mate4st**
**What this PR does / why we need it**:

Updates all CSI components for vsphere to version 2.7.0
This includes Snapshot controller.

In addition the namespace for the csi driver to work with all its features is required to be `vmware-system-csi`. This PR migrated the csi driver component to its desired namespace.

Usually we remove obsolete addons via `kubectl --prune` however this does not work between namespaces. Therefore a manual deletion of the resources in the old namespace is needed.

Note that `--prune` is an alpha/beta command. More information here:
https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands

**Which issue(s) this PR fixes**:

Fixes #10719

**What type of PR is this?**

/kind cleanup
/kind chore

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change? Then add your Release Note here**:

```release-note
Update vSphere CSI driver to v2.7.0. In addition the csi driver components have been moved to to their desired namespace.
```

**Documentation**:

```documentation
None
```

### CODE REVIEW

File: addons/csi/vsphere/csi-snapshot-validatingwebhook.yaml [hunk 1]
```diff
@@ -61,7 +60,7 @@ spec:
           type: RuntimeDefault
       containers:
         - name: snapshot-validation
-          image: {{ Image "registry.k8s.io/sig-storage/snapshot-validation-webhook:v5.0.1" }} # change the image if you wish to use your own custom validation server image
+          image: {{ Image "k8s.gcr.io/sig-storage/snapshot-validation-webhook:v6.0.1" }} # change the image if you wish to use your own custom validation server image
```
Comment (xmudrii) [inline addons/csi/vsphere/csi-snapshot-validatingwebhook.yaml]: Please use `registry.k8s.io` instead.

File: addons/csi/vsphere/setup-snapshot-controller.yaml [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+      labels:
+        app: snapshot-controller
+    spec:
+      serviceAccountName: snapshot-controller
+      securityContext:
+        seccompProfile:
+          type: RuntimeDefault
+      containers:
+        - name: snapshot-controller
+          image: {{ Image "gcr.io/k8s-staging-sig-storage/snapshot-controller:v6.0.1" }}
```
Comment (xmudrii) [inline addons/csi/vsphere/setup-snapshot-controller.yaml]: This should be `registry.k8s.io/sig-storage/snapshot-controller:v6.0.1`.

File: pkg/controller/seed-controller-manager/addon/addon_controller.go [hunk 3]
```diff
@@ -264,6 +263,21 @@ func (r *Reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, addo
...
+	if cluster.Spec.Cloud.VSphere != nil && addon.Name == "csi" {
+		legacyAddon := &kubermaticv1.Addon{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "internal-delete",
+			},
+			Spec: kubermaticv1.AddonSpec{
+				Name: "internal-delete",
+			},
+		}
+		if err := r.cleanupManifests(ctx, log, legacyAddon, cluster); err != nil {
```
Comment (xmudrii) [inline pkg/controller/seed-controller-manager/addon/addon_controller.go:276]: Is this going to error if it's already deleted or wasn't deployed in the first place?

File: pkg/resources/csi/vsphere/secret.go [hunk 4]
```diff
@@ -50,7 +51,7 @@ func CloudConfigSecretNameReconciler(data *resources.TemplateData) reconciling.N
 				return nil, err
 			}
 
-			cloudConfig, err := CloudConfigCSIToString(vsphereCloudConfig)
+			cloudConfig, err := vsphere.CloudConfigToString(vsphereCloudConfig)
```
Comment (xmudrii) [inline pkg/resources/csi/vsphere/secret.go]: Is this using the regular cloud-config? cloud-config for the CCM and the CSI is different, hence the `CloudConfigCSIToString` function. I don't think we should get rid of this function.

File: pkg/resources/resources.go [hunk 5]
```diff
@@ -865,17 +865,17 @@ const (
...
-	// CSIMigrationWebhookConfig is the name for the key that contains the webhook config.
-	CSIMigrationWebhookConfig = "webhook.config"
-	// CSIMigrationWebhookPort is the port used by the CSI-migration webhook.
-	CSIMigrationWebhookPort = 8443
-	// VsphereCSIMigrationWebhookConfigurationWebhookName is the webhook's name in the vSphere CSI_migration WebhookConfiguration.
-	VsphereCSIMigrationWebhookConfigurationWebhookName = "validation.csi.vsphere.vmware.com"
+	// VSphereCSINamespace is the namespace desired by vmware csi driver.
+	VSphereCSINamespace = "vmware-system-csi"
+	// VSphereCSIValidatingWebhookSVCName is the name of the validation webhook service.
+	VSphereCSIValidatingWebhookSVCName = "vsphere-webhook-svc"
```
Comment (xmudrii) [inline pkg/resources/resources.go]: The name is already long enough, I don't think we need to use `SVC`. :wink: 

```suggestion
	VSphereCSIValidatingWebhookServiceName = "vsphere-webhook-svc"
```

**Review by xmudrii — CHANGES_REQUESTED** [review]
Please replace all occurrences of `k8s.gcr.io` with `registry.k8s.io`.

File: pkg/controller/seed-controller-manager/addon/addon_controller.go (hunk 3 above)
Comment (mate4st) [reply pkg/controller/seed-controller-manager/addon/addon_controller.go:276]: It uses kubectl under the hood with `--ignore-not-found` as parameter.

File: pkg/resources/csi/vsphere/secret.go (hunk 4 above)
Comment (mate4st) [reply pkg/resources/csi/vsphere/secret.go]: Your are probably right with this one. It makes more sense to unify this in another PR as well. Right now we have multiple places where we maintain the conversion.

File: pkg/controller/user-cluster-controller-manager/resources/resources/csi/migration/secret.go [hunk 6]
```diff
@@ -24,38 +24,37 @@ import (
...
-port = "%d"
-cert-file = "/etc/webhook/cert.pem"
-key-file = "/etc/webhook/key.pem"
-`, resources.CSIMigrationWebhookPort)
+port = "8443"
+cert-file = "/run/secrets/tls/cert.pem"
+key-file = "/run/secrets/tls/key.pem"
+`
 
 func TLSServingCertificateReconciler(ca *triple.KeyPair) reconciling.NamedSecretReconcilerFactory {
```
Comment (xmudrii) [inline pkg/controller/user-cluster-controller-manager/resources/resources/csi/migration/secret.go]: Nit: can we name this function something like `VsphereTLSServingCertificateReconciler` to make it clear it's supposed to be only used for vSphere?

File: pkg/controller/user-cluster-controller-manager/resources/resources/csi/migration/webhook.go [hunk 7]
```diff
@@ -35,7 +35,8 @@ func ValidatingwebhookConfigurationReconciler(caCert *x509.Certificate, namespac
 			sideEffect := admissionregistrationv1.SideEffectClassNone
```
Comment (xmudrii) [inline pkg/controller/user-cluster-controller-manager/resources/resources/csi/migration/webhook.go:35]: Nit: I can't leave a comment on L32, but can we name this function something like `VsphereValidatingWebhookConfigurationReconciler` to make it clear it's supposed to be only used for vSphere?

**Comment by xmudrii** [comment]
@WeirdMachine LGTM, but please fix the verify job.

**Comment by xrstf** [comment]
@xmudrii PTALA

**Review by xmudrii — APPROVED** [review]
/lgtm
/approve
/hold
for manual testing

File: pkg/controller/seed-controller-manager/addon/addon_controller.go [hunk 8]
```diff
@@ -264,6 +263,21 @@ func (r *Reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, addo
 		return reqeueAfter, nil
 	}
 
+	// This handles manifests that should be removed manually, because kubectl --prune does not always delete obsolete resources.
```
Comment (xrstf) [inline pkg/controller/seed-controller-manager/addon/addon_controller.go:266]: Please explain more why this is here. Is kubectl broken? What is this "not always"?

/lgtm cancel

File: pkg/controller/seed-controller-manager/addon/addon_controller.go [hunk 9]
```diff
@@ -264,6 +263,21 @@ func (r *Reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, addo
...
+	if cluster.Spec.Cloud.VSphere != nil && addon.Name == "csi" {
+		legacyAddon := &kubermaticv1.Addon{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "internal-delete",
+			},
+			Spec: kubermaticv1.AddonSpec{
+				Name: "internal-delete",
```
Comment (xrstf) [inline pkg/controller/seed-controller-manager/addon/addon_controller.go:273]: I do not understand what is happening here... Why are we deleting an `internal-delete` addon when the CSI addon is being deleted?

**Comment by WeirdMachine** [comment]
Looking at this PR i am ashamed of myself. Therefore the csi driver is updated here: https://github.com/kubermatic/kubermatic/pull/11724

And the namespace migration should be done in another way.

**Pull request closed without merging**